editor/store/selectors.js

```diff
@@ -364,6 +364,46 @@ export function isEditedPostSaveable( state ) {
 	);
 }
 
+/**
+ * Returns true if the post can be autosaved, or false otherwise.
+ *
+ * @param  {Object}  state Global application state
+ * @return {Boolean}       Whether the post can be autosaved
+ */
+export function isPostAutosavable( state ) {
+
+	// If the post is autosaving, it is not autosavable.
+	if ( state.isAutosaving ) {
+		return false;
+	}
+
+	// If we don't already have an autosave, the post is autosavable.
+	if ( ! hasAutosave( state ) ) {
+		return true;
+	}
+
+	const title = getEditedPostTitle( state );
+	const excerpt = getEditedPostExcerpt( state );
+	const content = getEditedPostContent( state );
```

Comment:
yes, I agree - this needs refactoring.

The idea here is that once an autosave fires, another should not fire until you change something in a block. So this should function _similar to 'isDirty'_, working with the last autosaved data vs the last regular save. I couldn't figure out `state.editor.isDirty` gets set in the reducer - I think this should work in a similar way, can you tell me how that works?

Reply:
mmm nice idea, this works with a "higher order reducer" which means a function wrapper the editor's reducer and when we detect that the state instance has changed (since the state are immutable, the only way to change it is to generate a new instance), it means isDirty is true, we can do the same with `isAutosaveDirty`, the only difference is that the `isAutosaveDirty` should be reset when the autosave triggers.

It's done using the `withChangeDetection` Higher order reducer, we should make the `isDirty` key name customizable to be able to use it multiple times https://github.com/WordPress/gutenberg/blob/master/editor/store/reducer.js#L33

Reply:
ok, thanks for the explanation - i'll take another look at this.